Sources/PackageCollections/JSONModel/JSONCollection+v1.swift

```diff
@@ -241,3 +242,116 @@ extension JSONPackageCollectionModel.V1 {
         }
     }
 }
+
+// MARK: - Validations
+
+extension JSONPackageCollectionModel.V1 {
+    public struct Validator {
+        public let configuration: Configuration
+        
+        public init(configuration: Configuration = .init()) {
+            self.configuration = configuration
+        }
+        
+        public func validate(collection: Collection) -> [ValidationMessage]? {
+            var messages = [ValidationMessage]()
+            
+            let packages = collection.packages
+            // Stop validating if collection doesn't pass basic checks
+            if packages.isEmpty {
+                messages.append(.error("A collection must contain at least one package.", property: "packages"))
+            } else if packages.count > self.configuration.maximumPackageCount {
+                messages.append(.warning("The collection has (\(packages.count)) packages, which is more than the recommended maximum (\(self.configuration.maximumPackageCount)) and extra data might be ignored.", property: "packages"))
+            } else {
+                packages.forEach { self.validate(package: $0, messages: &messages) }
+            }
+            
+            guard messages.isEmpty else {
+                return messages
+            }
+            
+            return nil
+        }
+        
+        // TODO: validate package url?
+        private func validate(package: Collection.Package, messages: inout [ValidationMessage]) {
+            let packageID = package.url.absoluteString
```

Comment:
we may be able to reuse the new PackageIdentity type here which will help normalize across URL styles

Reply:
amended https://github.com/apple/swift-package-manager/pull/3076/commits/a9466edf4353cfd38a038d5595fd9875a21b84ec